lib/demuxer.ex

```diff
@@ -0,0 +1,160 @@
+    |> demand_if_not_blocked()
+  end
+
+  @impl true
+  def handle_demand(Pad.ref(:output, _id), _size, :buffers, _context, state) do
+    {:ok, state}
+  end
+
+  defp process_packets(packets_list, context, state) do
+    {actions, _context, state} =
+      Enum.reduce(packets_list, {[], context, state}, &process_packet/2)
+
+    {actions, state}
+  end
+
+  defp process_packet(packet, {actions, context, state}) do
+    cond do
+      packet.bos? -> process_bos_packet(packet, {actions, context, state})
+      packet.eos? and packet.payload == <<>> -> {actions, context, state}
+      true -> process_data_packet(packet, {actions, context, state})
+    end
+  end
+
+  defp process_bos_packet(
+         %{payload: <<"OpusHead", _rest::binary>>} = packet,
+         {actions, context, state}
+       ) do
+    new_track_action = {:notify_parent, {:new_track, {packet.track_id, :opus}}}
+    {actions ++ [new_track_action], context, %State{state | phase: :awaiting_linking}}
+  end
+
+  defp process_bos_packet(_other, _params) do
+    raise "Invalid bos packet, probably unsupported codec."
+  end
+
+  defp process_data_packet(%{payload: <<"OpusTags", _rest::binary>>}, {actions, context, state}) do
+    {actions, context, state}
+  end
+
+  defp process_data_packet(packet, {actions, context, state}) do
+    buffer_action =
+      {:buffer,
+       {Pad.ref(:output, packet.track_id),
+        %Buffer{
+          payload: packet.payload
+        }}}
+
+    classify_buffer_action(buffer_action, {actions, context, state})
+  end
+
+  defp demand_if_not_blocked({actions, state}) do
+    if blocked?(state) do
+      {actions, state}
+    else
+      actions = actions ++ [{:demand, :input}]
+      {actions, state}
+    end
+  end
+
+  defp blocked?(state) do
+    not Enum.empty?(state.actions_buffer) or state.phase != :all_outputs_linked
+  end
+
+  defp classify_buffer_action(
+         {:buffer, {Pad.ref(:output, id), _buffer}} = buffer_action,
+         {actions, context, state}
+       ) do
+    if not blocked?(state) and context.pads[Pad.ref(:output, id)].demand > 0 do
+      context = update_in(context.pads[Pad.ref(:output, id)].demand, &(&1 - 1))
+      {actions ++ [buffer_action], context, state}
+    else
+      {actions, context, %State{state | actions_buffer: state.actions_buffer ++ [buffer_action]}}
+    end
+  end
+
+  defp reclassify_buffer_actions({actions, state}, context) do
+    {actions, _context, state} =
+      Enum.reduce(
+        state.actions_buffer,
+        {actions, context, %State{state | actions_buffer: []}},
+        &classify_buffer_action/2
+      )
+
+    {actions, state}
+  end
+
+  @impl true
+  def handle_end_of_stream(:input, context, state) do
+    end_actions =
+      context.pads
+      |> Enum.filter(fn {_pad_ref, pad_data} -> pad_data.direction == :output end)
+      |> Enum.map(fn {pad_ref, _pad_data} -> {:end_of_stream, pad_ref} end)
+
+    {state.actions_buffer ++ end_actions, %State{state | actions_buffer: []}}
```

Comment:
There is forward action that will do it for you :wink: 

https://hexdocs.pm/membrane_core/Membrane.Element.Action.html#t:forward_t/0

```suggestion
    {state.actions_buffer ++ [forward: :end_of_stream], %State{state | actions_buffer: []}}
```